### ISSUE TITLE
Nodegroup is in transitional state "DELETE_FAILED"

### ISSUE THREAD

**Issue by arjun921** [weaveworks/eksctl#2216]
**What happened?**
A description of actual behavior (with error messages).
I was trying to delete few old nodegroups but I ran into the error listed below. 
```
➜  ~ eksctl delete nodegroup 1gpu-spot-p2-xlarge-p3-2xlarge --cluster=$ENVIRONMENT
[ℹ]  eksctl version 0.19.0
[ℹ]  using region us-east-1
[ℹ]  combined include rules: 1gpu-spot-p2-xlarge-p3-2xlarge
[ℹ]  1 nodegroup (1gpu-spot-p2-xlarge-p3-2xlarge) was included (based on the include/exclude rules)
[ℹ]  will delete 1 nodegroups from auth ConfigMap in cluster "kube-us-east-1"
[!]  error getting instance role ARN for nodegroup "1gpu-spot-p2-xlarge-p3-2xlarge": nodegroup "1gpu-spot-p2-xlarge-p3-2xlarge" is in transitional state ("DELETE_FAILED")                                                                                                             
```

I can't find the corresponding asg in EC2 so it seems to be deleted but still shows up when running 
`eksctl get nodegroup`.
**What you expected to happen?**
Expected behavior: nodegroup deletion should succeed.

**How to reproduce it?**
1. Create an eks cluster with following cluster.yml
```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: kube-us-east-1
  region: us-east-1
  version: '1.14'
  tags:
    environment: pre-production

nodeGroups:
  - name: ng-1
    desiredCapacity: 1
    minSize: 0
    maxSize: 10
    volumeSize: 100
    volumeType: gp2
    instanceType: m5a.2xlarge
    availabilityZones: ["us-east-1a"]
    labels:
      node-class: "worker-node"
    tags:
      # EC2 tags required for cluster-autoscaler auto-discovery
      k8s.io/cluster-autoscaler/enabled: "true"
      k8s.io/cluster-autoscaler/kube-us-east-1: "owned"
    iam:
      withAddonPolicies:
        albIngress: true
        autoScaler: true
        cloudWatch: true

  - name: ng-2
    desiredCapacity: 0
    minSize: 0
    maxSize: 10
    volumeSize: 20
    volumeType: gp2
    instanceType: m5a.2xlarge
    availabilityZones: ["us-east-1a"]
    labels:
      node-class: "worker-node"
    tags:
      # EC2 tags required for cluster-autoscaler auto-discovery
      k8s.io/cluster-autoscaler/enabled: "true"
      k8s.io/cluster-autoscaler/kube-us-east-1: "owned"
    iam:
      withAddonPolicies:
        albIngress: true
        autoScaler: true
        cloudWatch: true

  - name: 1gpu-spot-p2-xlarge-p3-2xlarge
    minSize: 0
    maxSize: 10
    instancesDistribution:
      maxPrice: 1.2
      instanceTypes: ["p2.xlarge","p3.2xlarge"]
      onDemandBaseCapacity: 0
      onDemandPercentageAboveBaseCapacity: 0
      spotAllocationStrategy: capacity-optimized
    labels:
      lifecycle: Ec2Spot
      aws.amazon.com/spot: "true"
      gpu-count: "1"
    availabilityZones: ["us-east-1a", "us-east-1b",  "us-east-1d",  "us-east-1f"]
    taints:
      spotInstance: "true:PreferNoSchedule"
    tags:
      k8s.io/cluster-autoscaler/node-template/label/lifecycle: Ec2Spot
      k8s.io/cluster-autoscaler/node-template/label/aws.amazon.com/spot: "true"
      k8s.io/cluster-autoscaler/node-template/label/gpu-count: "1"
      k8s.io/cluster-autoscaler/node-template/taint/spotInstance: "true:PreferNoSchedule"
      k8s.io/cluster-autoscaler/enabled: "true"
      k8s.io/cluster-autoscaler/kube-us-east-1: "owned"
    iam:
      withAddonPolicies:
        autoScaler: true
        cloudWatch: true
        albIngress: true

  - name: 4gpu-spot-p3-8xlarge
    minSize: 0
    maxSize: 4
    instancesDistribution:
      maxPrice: 4.4
      instanceTypes: ["p3.8xlarge"]
      onDemandBaseCapacity: 0
      onDemandPercentageAboveBaseCapacity: 0
      spotAllocationStrategy: capacity-optimized
    labels:
      lifecycle: Ec2Spot
      aws.amazon.com/spot: "true"
      gpu-count: "4"
    availabilityZones: ["us-east-1a", "us-east-1b",  "us-east-1d",  "us-east-1f"]
    taints:
      spotInstance: "true:PreferNoSchedule"
    tags:
      k8s.io/cluster-autoscaler/node-template/label/lifecycle: Ec2Spot
      k8s.io/cluster-autoscaler/node-template/label/aws.amazon.com/spot: "true"
      k8s.io/cluster-autoscaler/node-template/label/gpu-count: "4"
      k8s.io/cluster-autoscaler/node-template/taint/spotInstance: "true:PreferNoSchedule"
      k8s.io/cluster-autoscaler/enabled: "true"
      k8s.io/cluster-autoscaler/kube-us-east-1: "owned"
    iam:
      withAddonPolicies:
        autoScaler: true
        cloudWatch: true
        albIngress: true

  - name: 8gpu-spot-p3dn-24xlarge
    minSize: 0
    maxSize: 1
    instancesDistribution:
      maxPrice: 11
      instanceTypes: ["p3dn.24xlarge"]
      onDemandBaseCapacity: 0
      onDemandPercentageAboveBaseCapacity: 0
      spotAllocationStrategy: capacity-optimized
    labels:
      lifecycle: Ec2Spot
      aws.amazon.com/spot: "true"
      gpu-count: "8"
    availabilityZones: ["us-east-1a","us-east-1d"]
    taints:
      spotInstance: "true:PreferNoSchedule"
    tags:
      k8s.io/cluster-autoscaler/node-template/label/lifecycle: Ec2Spot
      k8s.io/cluster-autoscaler/node-template/label/aws.amazon.com/spot: "true"
      k8s.io/cluster-autoscaler/node-template/label/gpu-count: "8"
      k8s.io/cluster-autoscaler/node-template/taint/spotInstance: "true:PreferNoSchedule"
      k8s.io/cluster-autoscaler/enabled: "true"
      k8s.io/cluster-autoscaler/kube-us-east-1: "owned"
    iam:
      withAddonPolicies:
        autoScaler: true
        cloudWatch: true
        albIngress: true


availabilityZones: ["us-east-1a", "us-east-1b",  "us-east-1d",  "us-east-1f"]

```
2. Delete nodegroup 1gpu-spot-p2-xlarge-p3-2xlarge,4gpu-spot-p3-8xlarge or 8gpu-spot-p3dn-24xlarge

**Anything else we need to know?**
What OS are you using, are you using a downloaded binary or did you compile eksctl, what type of AWS credentials are you using (i.e. default/named profile, MFA) - please don't include actual credentials though!
OS: Kubuntu 20.04
AWS Credentials: Named Profile, with the following permissions:
![Screenshot_20200521_112416](https://user-images.githubusercontent.com/10341039/82528115-b20ddd80-9b55-11ea-8e89-92eb8ad84ab3.png)

**Versions**
```
➜  REDACTED git:(master) ✗ eksctl version
0.19.0
➜  REDACTED git:(master) ✗ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-11T18:14:22Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-f459c0", GitCommit:"f459c0672169dd35e77af56c24556530a05e9ab1", GitTreeState:"clean", BuildDate:"2020-03-18T04:24:17Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

**Logs**
https://gist.github.com/arjun921/30d2e79289b6796d87906b3bcf4d95d7

**Comment by michaelbeaumont** [comment]
@arjun921 I checked out the logs and I can see
```
StackStatusReason: "The following resource(s) failed to delete: [NodeInstanceRole]. ",  
```

If you look at the Cloudformation Stack `eksctl-kube-us-east-1-nodegroup-1gpu-spot-p2-xlarge-p3-2xlarge` in AWS and go to the Events, you should see one with `Logical ID` `NodeInstanceRole` and under `Status reason` some additional information about why it failed to delete. Could you check for me?

**Comment by pandvan** [comment]
I'm facing the same issue.
After trying to delete a nodegroup I've running the same problems described in #1325 and #1849 with the following output from eksctl:
```
[✖]  unexpected status "DELETE_FAILED" while waiting for CloudFormation stack "eksctl-k8s-staging-nodegroup-worker-1"
[ℹ]  fetching stack events in attempt to troubleshoot the root cause of the failure
[✖]  AWS::CloudFormation::Stack/eksctl-k8s-staging-nodegroup-worker-1: DELETE_FAILED – "The following resource(s) failed to delete: [NodeInstanceRole, SG]. "
[✖]  AWS::EC2::SecurityGroup/SG: DELETE_FAILED – "resource sg-0ad4902c75142e3d6 has a dependent object (Service: AmazonEC2; Status Code: 400; Error Code: DependencyViolation; Request ID: 1e730987-d2f3-4e5b-87de-456e61434448)"
[✖]  AWS::IAM::Role/NodeInstanceRole: DELETE_FAILED – "Cannot delete entity, must detach all policies first. (Service: AmazonIdentityManagement; Status Code: 409; Error Code: DeleteConflict; Request ID: 2a9c09ee-b5e0-47d0-a753-475b73c704f1)"
[ℹ]  1 error(s) occurred while deleting nodegroup(s)
[✖]  waiting for CloudFormation stack "eksctl-k8s-staging-nodegroup-worker-1": ResourceNotReady: failed waiting for successful resource state
```
After that I've manually deleted the secure group as indicated by other users in the mentioned issues. When I re-run eksctl to delete the node group using `eksctl delete nodegroup -f cluster_staging.yml --wait --only-missing --approve` it fails with ` nodegroup "worker-1" is in transitional state ("DELETE_FAILED")`

**Versions**
```
$ eksctl version
0.20.0
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.8", GitCommit:"ec6eb119b81be488b030e849b9e64fda4caaf33c", GitTreeState:"clean", BuildDate:"2020-03-12T21:00:06Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.8-eks-e16311", GitCommit:"e163110a04dcb2f39c3325af96d019b4925419eb", GitTreeState:"clean", BuildDate:"2020-03-27T22:37:12Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by michaelbeaumont** [comment]
@pandvan It can be hard to reproduce these deletion errors. Would you be willing to try rerunning `delete nodegroup` with my branch [michaelbeaumont/eksctl#ng_delete](https://github.com/michaelbeaumont/eksctl/tree/ng_delete)?

**Comment by pandvan** [comment]
@michaelbeaumont I'm very sorry but I had to go through and delete the nodegroup stack directly from CloudFormation as it happend un a pre-production cluster.
I've found 2 major stop points that prevent eksctl/CloudFormation to successfully delete the nodegroup:
1.  There were some ENI attached to nodegroup local secure group due to AWS EFS mountpoints that refer tp that secure group
2.  There was a policy still associated to the NodeInstanceRole, manually added following AWS guide to let eks cluster work with EBS CSI driver. I don't think this was really helpful as ekcstl takes care of this on his own.

After manually remote the policy and the local (not shared) nodegrou's SG I was able to perfom the deletion from CloudFormation and then ekcsctl corectly detect that nodegroup does no longer exist.

Thanks for your prompt help and sorry again if not being so helpful to further investigation.
Anyway, is there any way to force eksctl to retry CloudFormation stack deletion event if it was failed before?
